File: backend/app/routers/theme.py
"""Tema personalizado (protótipo) — o coletor só guarda os bytes que o painel
manda e devolve pra placa buscar (GET). Não valida o JSON do tema (é opaco
pro coletor, quem entende o schema é o firmware/frontend — ver
.agents/CONTRATO_TEMA.md); não faz parte do contrato de /usage."""

from __future__ import annotations

from fastapi import APIRouter, HTTPException, Request
from fastapi.responses import Response

from app.config import data_dir

router = APIRouter(prefix="/api/theme", tags=["theme"])

_MAX_META_BYTES = 8192


def _meta_path():
    return data_dir() / "theme.json"


def _wallpapers_dir():
    return data_dir() / "wallpapers"


def _wallpaper_raw_path(wid: str, suffix: str = ""):
    if suffix:
        return _wallpapers_dir() / f"{wid}{suffix}.raw"
    return _wallpapers_dir() / f"{wid}.raw"


def _screen_suffix(request: Request) -> str:
    screen = request.headers.get("X-Vigia-Screen") or request.headers.get("x-vigia-screen") or ""
    suffix = ""
    if screen and "x" in screen:
        try:
            ws, _, hs = screen.partition("x")
            wi = int(ws)
            hi = int(hs)
            if wi == 320 and hi == 240:
                suffix = "_wokwi"
        except ValueError:
            pass
    w = request.query_params.get("w")
    h = request.query_params.get("h")
    if w and h:
        try:
            wi = int(w)
            hi = int(h)
            if wi == 160 and hi == 120:
                suffix = "_wokwi"
            elif wi == 240 and hi == 160:
                suffix = ""
        except ValueError:
            pass
    return suffix


@router.get(
    "",
    summary="Tema salvo (protótipo)",
    description="A placa busca aqui (botão de recarregar no header) pra aplicar o último tema que o painel salvou.",
)
def get_theme() -> dict:
    from app.routers.wallpapers import _get_selected_id

    meta_path = _meta_path()
    theme = None
    if meta_path.is_file():
        theme = meta_path.read_text(encoding="utf-8")
    selected_id = _get_selected_id()
    return {
        "active": theme is not None,
        "theme": theme,
        "has_background": selected_id is not None,
        "background_id": selected_id,
    }


@router.post("/meta", summary="Salva o theme.json (protótipo)")
async def post_theme_meta(request: Request) -> dict:
    body = await request.body()
    if not body:
        raise HTTPException(400, "corpo vazio")
    if len(body) > _MAX_META_BYTES:
        raise HTTPException(413, "tema grande demais")
    data_dir().mkdir(parents=True, exist_ok=True)
    _meta_path().write_bytes(body)
    return {"ok": True}


@router.get(
    "/background",
    summary="Bytes RAW do papel de parede selecionado",
    response_class=Response,
)
def get_theme_background(request: Request) -> Response:
    from app.routers.wallpapers import _get_selected_id, _image_to_raw

    wid = _get_selected_id()
    if not wid:
        raise HTTPException(404, "nenhum papel de parede selecionado")
    suffix = _screen_suffix(request)
    wp_path = _wallpaper_raw_path(wid, suffix)
    if wp_path.is_file():
        return Response(content=wp_path.read_bytes(), media_type="application/octet-stream")
    wp_path2 = _wallpaper_raw_path(wid)
    if wp_path2.is_file():
        return Response(content=wp_path2.read_bytes(), media_type="application/octet-stream")
    orig = _wallpapers_dir() / f"{wid}.orig"
    if orig.is_file():
        try:
            tw, th = (160, 120) if suffix == "_wokwi" else (240, 160)
            raw = _image_to_raw(orig.read_bytes(), tw, th)
            return Response(content=raw, media_type="application/octet-stream")
        except Exception:
            pass
    raise HTTPException(404, "papel de parede não encontrado")


@router.get("/background/index", include_in_schema=False)
def get_background_index() -> dict:
    """Compat: placas antigas polleavam o slideshow. Sempre desligado."""
    from app.routers.wallpapers import _get_selected_id

    wid = _get_selected_id()
    return {"enabled": False, "index": 0, "count": 1 if wid else 0, "interval": 0, "current_id": wid}


@router.delete("", summary="Apaga o tema salvo (protótipo)")
def delete_theme() -> dict:
    _meta_path().unlink(missing_ok=True)
    return {"ok": True}

File: backend/src/routers/theme.ts
import type { FastifyInstance } from "fastify";
import { existsSync, readFileSync, writeFileSync, mkdirSync, unlinkSync } from "node:fs";
import { join } from "node:path";
import { dataDir } from "../config.js";

const MAX_META_BYTES = 8192;

function metaPath(): string {
  return join(dataDir(), "theme.json");
}
function wallpapersDir(): string {
  return join(dataDir(), "wallpapers");
}
function wallpaperRawPath(wid: string, suffix = ""): string {
  if (suffix) return join(wallpapersDir(), `${wid}${suffix}.raw`);
  return join(wallpapersDir(), `${wid}.raw`);
}

function screenSuffix(request: unknown): string {
  const req = request as { headers: Record<string, string | undefined>; query: Record<string, string | undefined> };
  const headers = req.headers ?? {};
  const screen = headers["x-vigia-screen"] ?? headers["X-Vigia-Screen"] ?? "";
  let suffix = "";
  if (screen && screen.includes("x")) {
    try {
      const [ws, , hs] = partition(screen, "x");
      const wi = parseInt(ws, 10);
      const hi = parseInt(hs, 10);
      if (wi === 320 && hi === 240) suffix = "_wokwi";
    } catch {}
  }
  const w = req.query?.w;
  const h = req.query?.h;
  if (w && h) {
    try {
      const wi = parseInt(w, 10);
      const hi = parseInt(h, 10);
      if (wi === 160 && hi === 120) suffix = "_wokwi";
      else if (wi === 240 && hi === 160) suffix = "";
    } catch {}
  }
  return suffix;
}

function partition(s: string, sep: string): [string, string, string] {
  const idx = s.indexOf(sep);
  if (idx === -1) return [s, "", ""];
  return [s.slice(0, idx), sep, s.slice(idx + sep.length)];
}

export async function createThemeRoutes(app: FastifyInstance): Promise<void> {
  app.get("/api/theme", async () => {
    let theme: string | null = null;
    const p = metaPath();
    if (existsSync(p)) {
      try {
        theme = readFileSync(p, "utf-8");
      } catch {}
    }
    // get selected id via wallpapers helper if available
    let selectedId: string | null = null;
    try {
      const wallpapers = await import("./wallpapers/router.js");
      // wallpapers module may expose _getSelectedId equivalent; try to call internal helper via reading config
      const { load } = await import("../store.js");
      const cfg = load() as Record<string, unknown>;
      const wp = (cfg.wallpapers as Record<string, unknown>) ?? {};
      selectedId = String(wp.selected_id ?? "").trim() || null;
      // verify file exists else null
      if (selectedId) {
        // check if wallpaper exists via meta
        const { existsSync: es } = await import("node:fs");
        const { join: j } = await import("node:path");
        const { dataDir: dd } = await import("../config.js");
        const metaFile = j(dd(), "wallpapers.json");
        if (es(metaFile)) {
          try {
            const raw = JSON.parse(readFileSync(metaFile, "utf-8")) as Record<string, unknown>;
            const list = (raw.wallpapers ?? []) as Array<Record<string, unknown>>;
            const found = list.some((w) => String(w.id) === selectedId);
            if (!found) {
              // fallback: check raw file existence
              const rawPath = j(dd(), "wallpapers", `${selectedId}.raw`);
              if (!es(rawPath) && !es(j(dd(), "wallpapers", `${selectedId}_wokwi.raw`))) selectedId = null;
            }
          } catch { selectedId = null; }
        }
      }
    } catch {}
    return {
      active: theme !== null,
      theme,
      has_background: selectedId !== null,
      background_id: selectedId,
    };
  });

  app.post("/api/theme/meta", async (request, reply) => {
    const body = await getRawBody(request);
    if (!body || body.length === 0) return reply.code(400).send({ ok: false, error: "corpo vazio" });
    if (body.length > MAX_META_BYTES) return reply.code(413).send({ ok: false, error: "tema grande demais" });
    try {
      mkdirSync(dataDir(), { recursive: true });
      const tmp = metaPath() + ".tmp";
      writeFileSync(tmp, body);
      const { renameSync, chmodSync } = await import("node:fs");
      renameSync(tmp, metaPath());
      try { chmodSync(metaPath(), 0o600); } catch {}
    } catch (e) {
      return reply.code(500).send({ ok: false, error: String(e) });
    }
    return { ok: true };
  });

  app.get("/api/theme/background", async (request, reply) => {
    // dynamic import wallpapers helpers
    const wid = await resolveSelectedId();
    if (!wid) return reply.code(404).send({ ok: false, error: "nenhum papel de parede selecionado" });
    const suffix = screenSuffix(request);
    let p = wallpaperRawPath(wid, suffix);
    if (existsSync(p)) {
      const data = readFileSync(p);
      return reply.type("application/octet-stream").send(data);
    }
    p = wallpaperRawPath(wid);
    if (existsSync(p)) {
      const data = readFileSync(p);
      return reply.type("application/octet-stream").send(data);
    }
    const orig = join(wallpapersDir(), `${wid}.orig`);
    if (existsSync(orig)) {
      try {
        const bytes = readFileSync(orig);
        const { imageToRaw } = await import("./wallpapers/router.js");
        const tw = suffix === "_wokwi" ? 160 : 240;
        const th = suffix === "_wokwi" ? 120 : 160;
        const raw = await (imageToRaw as (b: Buffer, w: number, h: number) => Promise<Buffer>)(bytes, tw, th);
        return reply.type("application/octet-stream").send(raw);
      } catch {}
    }
    return reply.code(404).send({ ok: false, error: "papel de parede não encontrado" });
  });

  app.get("/api/theme/background/index", async () => {
    const wid = await resolveSelectedId();
    return { enabled: false, index: 0, count: wid ? 1 : 0, interval: 0, current_id: wid };
  });

  app.delete("/api/theme", async () => {
    try { unlinkSync(metaPath()); } catch {}
    return { ok: true };
  });
}

async function resolveSelectedId(): Promise<string | null> {
  try {
    const { load } = await import("../store.js");
    const cfg = load() as Record<string, unknown>;
    const wp = (cfg.wallpapers as Record<string, unknown>) ?? {};
    const selected = String(wp.selected_id ?? "").trim();
    if (!selected) return null;
    // validate existence via wallpapers.json or raw
    const { existsSync } = await import("node:fs");
    const { join } = await import("node:path");
    const { dataDir: dd } = await import("../config.js");
    const raw1 = join(dd(), "wallpapers", `${selected}.raw`);
    const raw2 = join(dd(), "wallpapers", `${selected}_wokwi.raw`);
    const preview = join(dd(), "wallpapers", `${selected}.jpg`);
    const orig = join(dd(), "wallpapers", `${selected}.orig`);
    if (existsSync(raw1) || existsSync(raw2) || existsSync(preview) || existsSync(orig)) return selected;
    // fallback to first wallpaper if any
    const metaPath2 = join(dd(), "wallpapers.json");
    if (existsSync(metaPath2)) {
      try {
        const data = JSON.parse(readFileSync(metaPath2, "utf-8")) as Record<string, unknown>;
        const list = (data.wallpapers ?? []) as Array<Record<string, unknown>>;
        if (list.length > 0) return String(list[0].id);
      } catch {}
    }
    return null;
  } catch { return null; }
}

async function getRawBody(request: unknown): Promise<Buffer> {
  const anyReq = request as unknown as { raw?: { read?: () => unknown }; body?: unknown };
  if (Buffer.isBuffer(anyReq.body)) return anyReq.body as Buffer;
  if (typeof anyReq.body === "string") return Buffer.from(anyReq.body as string, "utf-8");
  if (anyReq.body !== null && typeof anyReq.body === "object") {
    // Check if it's already a buffer-like from multipart? Return JSON string
    try { return Buffer.from(JSON.stringify(anyReq.body), "utf-8"); } catch { return Buffer.alloc(0); }
  }
  // Try reading raw stream
  try {
    const raw = (request as unknown as { raw?: { read: () => Buffer } }).raw;
    if (raw && typeof raw.read === "function") {
      const data = raw.read();
      if (Buffer.isBuffer(data) && data.length) return data;
    }
  } catch {}
  return Buffer.alloc(0);
}
